cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AbstractDistributedProgramRunner.java

```diff
@@ -157,19 +168,50 @@ protected EventHandler createEventHandler(CConfiguration cConf) {
     return new AbortOnTimeoutEventHandler(cConf.getLong(Constants.CFG_TWILL_NO_CONTAINER_TIMEOUT, Long.MAX_VALUE));
   }
 
+  // returns the path of the file, if its a local URI. If it's an HDFS URI, copy it to local file system first.
+  private String getKeytabPath(URI keytabURI, File tempDir) throws IOException {
+    if (!"hdfs".equals(keytabURI.getScheme())) {
+      // if it's not hdfs, assume its local file
+      if (keytabURI.getScheme() != null && !"file".equals(keytabURI.getScheme())) {
+        throw new IllegalArgumentException("Currently only file:// and hdfs:// URIs are supported.");
+      }
+      return keytabURI.getPath();
+    }
+
+    // create a local file with restricted permissions
+    File localKeytabFile = new File(tempDir, "keytab.localized");
```

Comment:
Use the new Java `Files.createFile` instead so that you can set the right permission when the file is created.